src/openms/include/OpenMS/ANALYSIS/ID/PeptideIndexing.h

```diff
@@ -677,8 +699,155 @@ namespace OpenMS
       return EXECUTION_OK;
     }
 
-protected:
-    struct PeptideProteinMatchInformation
+ protected:
+
+     using MapDecoyStringToPrefixSuffixCount = std::map<std::string, std::pair<int, int>>;
+     using MapCaseInsensitiveToCaseSensitiveDecoyString = std::map<std::string, std::string>;
+
+     template<typename T>
+     ExitCodes findDecoyString_(FASTAContainer<T> & proteins,
+                                std::string & decoy_string,
+                                bool & is_prefix)
+     {
+       // Common decoy strings in fasta files. Used to determine if a protein is target/decoy automatically.
+       // Note: decoy prefixes/suffices must be provided in lower case and must only contain alphanumeric characters
+        MapDecoyStringToPrefixSuffixCount decoy_count = {{"decoy", {0,0}}, {"rev", {0,0}},
+                                                          {"reverse", {0,0}}, {"dev", {0,0}},
+                                                          {"iddecoy", {0,0}}, {"xxx", {0,0}},
+                                                          {"shuffle", {0,0}}, {"shuffled", {0,0}},
+                                                          {"random", {0,0}}};
+
+
+       // map case insensitive strings back to original case (as used in fasta)
+       MapCaseInsensitiveToCaseSensitiveDecoyString decoy_case_sensitive;
+
+       const size_t PROTEIN_CACHE_SIZE = 4e5;
+
+       while (true)
+       {
+         proteins.cacheChunk(PROTEIN_CACHE_SIZE);
+         bool has_active_data = proteins.activateCache();
+
+         if (!has_active_data) break;
+         auto prot_count = (SignedSize)proteins.chunkSize();
+
+         {
+           for (SignedSize i = 0; i < prot_count; ++i)
+           {
+             String seq = proteins.chunkAt(i).identifier;
+             String seq_lower = seq;
+             seq_lower.toLower();
+
+             // remove all non-alphanumeric characters
+             seq_lower.erase(std::remove_if(seq_lower.begin(), seq_lower.end(),
```

Comment:
what is the rationale here?
Have you seen weird characters in the wild, which warrant this overhead?

Reply:
Input files seem to containt decoy strings with sometimes arbitrary amounts of '_' characters.
In order to compare things more easily I simply remove all non-alphanumeric characters. Since the identifiers are usually pretty small strings this should be fine?

Reply:
how would these underscore cases look like?
e.g. "rev____albumin" would be detected in any case. So there are people out there who use "_rev____albumin"?
If thats the case, ok.

Reply:
I have observed cases like:
__id_decoy__blabla